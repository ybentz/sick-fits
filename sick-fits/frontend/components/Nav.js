import React from 'react'
import Link from 'next/link'
import { useMutation } from '@apollo/react-hooks'
import NavStyles from './styles/NavStyles'
import { useUser } from './User'
import Signout from './Signout'
import { TOGGLE_CART_MUTATION } from './Cart'
import CartCount from './CartCount'

function Nav() {
  const user = useUser()
  const [toggleCart] = useMutation(TOGGLE_CART_MUTATION)

  return (
    <NavStyles>
      <li>
        <Link href="/items">
          <a>Shop</a>
        </Link>
      </li>
      {user && (
        <>
          <li>
            <Link href="/sell">
              <a>Sell</a>
            </Link>
          </li>
          <li>
            <Link href="/orders">
              <a>Orders</a>
            </Link>
          </li>
          <li>
            <Link href="/me">
              <a>Account</a>
            </Link>
          </li>
          <li>
            <Signout />
          </li>
          <li>
            <button onClick={toggleCart}>
              My Cart
              <CartCount
                count={user.cart.reduce((acc, item) => acc + item.quantity, 0)}
              />
            </button>
          </li>
        </>
      )}
      {!user && (
        <li>
          <Link href="/signup">
            <a>Signin</a>
          </Link>
        </li>
      )}
    </NavStyles>
  )
}

export default Nav
