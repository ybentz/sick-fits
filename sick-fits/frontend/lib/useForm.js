import { useState } from 'react'

function useForm(initial = {}) {
  const [inputs, setInputs] = useState(initial)

  const handleChange = (event) => {
    const { name, type, value } = event.target
    const val = type === 'number' ? parseFloat(value) : value
    setInputs({
      ...inputs,
      [name]: val,
    })
  }

  return {
    inputs,
    handleChange,
  }
}

export default useForm
