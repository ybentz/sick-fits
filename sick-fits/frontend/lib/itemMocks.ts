// import { MockedResponse } from '@apollo/react-testing'

import { fakeUser } from './testUtils'
import { SINGLE_ITEM_QUERY } from '../components/SingleItem'

class ItemMockBuilder implements Item {
  id: string
  description: string
  image: string
  largeImage: string
  price: number
  title: string
  user: User

  constructor() {
    this.id = '1234'
    this.description = 'Test description'
    this.image = 'test-dog-small.jpg'
    this.largeImage = 'test-dog.jpg'
    this.price = 5000
    this.title = 'Test Product Title'
    this.user = fakeUser()
  }
  withId(id: string) {
    this.id = id
    return this
  }
  withDescription(description: string) {
    this.description = description
    return this
  }
  withImage(url: string) {
    this.image = url
    return this
  }
  withLargeImage(url: string) {
    this.largeImage = url
    return this
  }
  withPrice(price: number) {
    this.price = price
    return this
  }
  withTitle(title: string) {
    this.title = title
    return this
  }
  withUser(user: User) {
    this.user = user
    return this
  }
  build(): Item {
    return {
      description: this.description,
      id: this.id,
      image: this.image,
      largeImage: this.largeImage,
      price: this.price,
      title: this.title,
      user: this.user,
    }
  }
}

class SingleItemQueryMockBuilder extends ItemMockBuilder {
  errorMessage: string

  withError(message: string) {
    this.errorMessage = message
    return this
  }

  // Should be returning type `MockedResponse` instead of `any but that complicates things
  // because TS doesn't handle the overload with same parameters signature this way
  build(): any {
    const request = { query: SINGLE_ITEM_QUERY, variables: { id: this.id } }
    if (this.errorMessage) {
      return {
        request,
        result: {
          errors: [{ message: this.errorMessage }],
        },
      }
    }

    return {
      request,
      result: {
        data: {
          item: {
            description: this.description,
            id: this.id,
            image: this.image,
            largeImage: this.largeImage,
            price: this.price,
            title: this.title,
            user: this.user,
          },
        },
      },
    }
  }
}

export { ItemMockBuilder, SingleItemQueryMockBuilder }
