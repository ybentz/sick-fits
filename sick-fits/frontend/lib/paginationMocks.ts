import { PAGINATION_QUERY } from '../components/Pagination'

class PaginationQueryMockBuilder {
  count: number
  errorMessage: string

  constructor() {
    this.count = 1
  }
  withItemCount(count: number) {
    this.count = count
    return this
  }
  withError(message: string) {
    this.errorMessage = message
    return this
  }

  build() {
    const request = { query: PAGINATION_QUERY }
    if (this.errorMessage) {
      return {
        request,
        result: {
          errors: [{ message: this.errorMessage }],
        },
      }
    }

    return {
      request,
      result: {
        data: {
          itemsConnection: {
            aggregate: {
              count: this.count,
            },
          },
        },
      },
    }
  }
}

export { PaginationQueryMockBuilder }
