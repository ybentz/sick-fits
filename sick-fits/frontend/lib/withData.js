import { InMemoryCache } from 'apollo-cache-inmemory'
import { ApolloClient } from 'apollo-client'
import { ApolloLink, Observable } from 'apollo-link'
import { onError } from 'apollo-link-error'
import { HttpLink } from 'apollo-link-http'
import withApollo from 'next-with-apollo'

import { LOCAL_STATE_QUERY } from '../components/Cart'
import { endpoint } from '../config'

const cache = new InMemoryCache()

async function request(operation, options) {
  operation.setContext({
    fetchOptions: {
      credentials: 'include',
    },
    headers: options.headers,
  })
}

function requestLinkWithOptions(options) {
  return new ApolloLink(
    (operation, forward) =>
      new Observable((observer) => {
        let handle
        Promise.resolve(operation)
          .then((oper) => request(oper, options))
          .then(() => {
            handle = forward(operation).subscribe({
              next: observer.next.bind(observer),
              error: observer.error.bind(observer),
              complete: observer.complete.bind(observer),
            })
          })
          .catch(observer.error.bind(observer))
        return () => {
          if (handle) handle.unsubscribe()
        }
      })
  )
}

function createClient({ headers }) {
  return new ApolloClient({
    cache,
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors)
          // Handle/log graphQL errors - just console logging for now
          graphQLErrors.forEach(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
          )
        // Handle/log network errors - just console logging for now
        if (networkError) console.log(`[Network error]: ${networkError}`)
      }),
      requestLinkWithOptions({ headers }),
      new HttpLink({
        uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
        credentials: 'include',
      }),
    ]),
    resolvers: {
      Mutation: {
        // first argument unknown and unused lol
        toggleCart(_, variables, { cache }) {
          // read cartOpen value from cache
          const { cartOpen } = cache.readQuery({
            query: LOCAL_STATE_QUERY,
          })
          const data = {
            data: { cartOpen: !cartOpen },
          }
          cache.writeData(data)
          return data
        },
      },
    },
  })
}

// Add default local state
cache.writeData({
  data: {
    cartOpen: false,
  },
})

export default withApollo(createClient)
