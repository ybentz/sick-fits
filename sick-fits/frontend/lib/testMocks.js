import { CURRENT_USER_QUERY } from '../components/User'
import { fakeUser, fakeCartItem } from '../lib/testUtils'

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: null } },
  },
]

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: fakeUser() } },
  },
]

const signedInMocksWithCartItems = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
        },
      },
    },
  },
]

class UserMockBuilder {
  constructor() {
    this.cart = []
    this.user = null
  }
  setSignedIn() {
    this.user = fakeUser()
    return this
  }
  setSignedOut() {
    this.user = null
    return this
  }
  withCartItems(count = 1, quantity = 1) {
    this.cart = Array(count).fill(fakeCartItem({ quantity }))
    return this
  }
  build() {
    return {
      request: { query: CURRENT_USER_QUERY },
      result: {
        data: {
          me: !this.user
            ? null
            : {
                ...this.user,
                cart: this.cart,
              },
        },
      },
    }
  }
}

export {
  notSignedInMocks,
  signedInMocks,
  signedInMocksWithCartItems,
  UserMockBuilder,
}
