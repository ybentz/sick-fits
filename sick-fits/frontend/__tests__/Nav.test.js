import { MockedProvider } from '@apollo/react-testing'
import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import React from 'react'

import Nav from '../components/Nav'
import { UserMockBuilder } from '../lib/testMocks'
import { waitForApolloStateChange } from '../lib/testUtils'

describe('<Nav/>', () => {
  it('renders minimal nav when sign out', async () => {
    const mockedUser = new UserMockBuilder().setSignedOut().build()
    const { asFragment } = render(
      <MockedProvider mocks={[mockedUser]} addTypename={false}>
        <Nav />
      </MockedProvider>
    )
    await waitForApolloStateChange()
    // await waitFor(() => {})
    expect(screen.getByRole('link', { name: /signin/i })).toBeInTheDocument()
    expect(screen.getByRole('link', { name: /shop/i })).toBeInTheDocument()

    expect(screen.queryByRole('link', { name: /sell/i })).toBeNull()
    expect(screen.queryByRole('link', { name: /orders/i })).toBeNull()
    expect(screen.queryByRole('link', { name: /account/i })).toBeNull()
    expect(screen.queryByRole('button', { name: /sign out/i })).toBeNull()
    expect(screen.queryByRole('button', { name: /my cart/i })).toBeNull()
    // expect(asFragment()).toMatchSnapshot()
  })

  it('renders full nav when signed in', async () => {
    const mockedUser = new UserMockBuilder().setSignedIn().build()
    const { asFragment } = render(
      <MockedProvider mocks={[mockedUser]} addTypename={false}>
        <Nav />
      </MockedProvider>
    )
    await waitForApolloStateChange()
    // await waitFor(() => {})
    expect(screen.getByRole('link', { name: /shop/i })).toBeInTheDocument()
    expect(screen.getByRole('link', { name: /sell/i })).toBeInTheDocument()
    expect(screen.getByRole('link', { name: /orders/i })).toBeInTheDocument()
    expect(screen.getByRole('link', { name: /account/i })).toBeInTheDocument()
    expect(
      screen.getByRole('button', { name: /sign out/i })
    ).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /my cart/i })).toBeInTheDocument()

    expect(screen.queryByRole('link', { name: /signin/i })).toBeNull()
    // expect(asFragment()).toMatchSnapshot()
  })

  it('renders the amount of items in the cart', async () => {
    const cartItemsCount = 4
    const mockedUser = new UserMockBuilder()
      .setSignedIn()
      .withCartItems(cartItemsCount)
      .build()

    const { asFragment } = render(
      <MockedProvider mocks={[mockedUser]} addTypename={false}>
        <Nav />
      </MockedProvider>
    )
    await waitForApolloStateChange()
    // await waitFor(() => {})
    const cartButton = await screen.findByRole('button', {
      name: /my cart/i,
    })
    expect(cartButton).toHaveTextContent(cartItemsCount)
    // expect(asFragment()).toMatchSnapshot()
  })
})
