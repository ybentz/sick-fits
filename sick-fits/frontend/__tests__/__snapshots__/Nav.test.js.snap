// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> renders full nav when signed in 1`] = `
<DocumentFragment>
  <ul
    class="NavStyles-sc-11c0d2g-0 iCTkso"
  >
    <a
      href="/items"
    >
      Shop
    </a>
    <a
      href="/sell"
    >
      Sell
    </a>
    <a
      href="/orders"
    >
      Orders
    </a>
    <a
      href="/me"
    >
      Account
    </a>
    <button>
      Sign Out
    </button>
    <button>
      My Cart
      <span
        class="CartCount__AnimationStyles-xxvp4g-0 jGgtAN"
      >
        <div>
          <div
            class="CartCount__Dot-xxvp4g-1 kCeHkh count"
          >
            0
          </div>
        </div>
      </span>
    </button>
  </ul>
</DocumentFragment>
`;

exports[`<Nav/> renders minimal nav when sign out 1`] = `
<DocumentFragment>
  <ul
    class="NavStyles-sc-11c0d2g-0 iCTkso"
  >
    <a
      href="/items"
    >
      Shop
    </a>
    <a
      href="/signup"
    >
      Signin
    </a>
  </ul>
</DocumentFragment>
`;

exports[`<Nav/> renders the amount of items in the cart 1`] = `
<DocumentFragment>
  <ul
    class="NavStyles-sc-11c0d2g-0 iCTkso"
  >
    <a
      href="/items"
    >
      Shop
    </a>
    <a
      href="/sell"
    >
      Sell
    </a>
    <a
      href="/orders"
    >
      Orders
    </a>
    <a
      href="/me"
    >
      Account
    </a>
    <button>
      Sign Out
    </button>
    <button>
      My Cart
      <span
        class="CartCount__AnimationStyles-xxvp4g-0 jGgtAN"
      >
        <div>
          <div
            class="CartCount__Dot-xxvp4g-1 kCeHkh count"
          >
            4
          </div>
        </div>
      </span>
    </button>
  </ul>
</DocumentFragment>
`;
